java-frontend/src/test/java/org/sonar/java/cfg/CFGLoopTest.java

```diff
@@ -0,0 +1,204 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.cfg;
+
+import org.junit.Test;
+import org.sonar.java.cfg.CFG.Block;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import java.util.List;
+import java.util.Map;
+
+import static org.fest.assertions.Assertions.assertThat;
+
+public class CFGLoopTest {
+
+  private static final CFGTestLoader loader = new CFGTestLoader("src/test/files/cfg/CFGLoopTest.java");
+
+  @Test
+  public void simpleWhileLoop() {
+    final CFG cfg = CFG.build(loader.getMethod("CFGLoopTest", "simpleWhileLoop"));
+    Map<Tree, CFGLoop> loops = CFGLoop.getCFGLoops(cfg);
+    assertThat(loops).hasSize(1);
+    CFGLoop loop = loops.values().iterator().next();
+    List<Block> blocks = CFGLoop.sorted(cfg.blocks());
+    assertThat(loop.startingBlock()).isSameAs(blocks.get(2));
+    assertThat(loop.blocks()).containsOnly(blocks.get(1));
+    assertThat(loop.hasNoWayOut()).isTrue();
+  }
+
+  @Test
+  public void simpleWhileLoopWithBreak() {
+    final CFG cfg = CFG.build(loader.getMethod("CFGLoopTest", "simpleWhileLoopWithBreak"));
+    Map<Tree, CFGLoop> loops = CFGLoop.getCFGLoops(cfg);
+    assertThat(loops).hasSize(1);
+    CFGLoop loop = loops.values().iterator().next();
+    List<Block> blocks = CFGLoop.sorted(cfg.blocks());
+    assertThat(loop.startingBlock()).isSameAs(blocks.get(3));
+    assertThat(loop.blocks()).containsOnly(blocks.get(2), blocks.get(1));
+    assertThat(loop.hasNoWayOut()).isFalse();
+  }
+
+  @Test
+  public void simpleForLoop() {
+    final CFG cfg = CFG.build(loader.getMethod("CFGLoopTest", "simpleForLoop"));
+    Map<Tree, CFGLoop> loops = CFGLoop.getCFGLoops(cfg);
+    assertThat(loops).hasSize(1);
+    CFGLoop loop = loops.values().iterator().next();
+    List<Block> blocks = CFGLoop.sorted(cfg.blocks());
+    assertThat(loop.startingBlock()).isSameAs(blocks.get(3));
+    assertThat(loop.blocks()).containsOnly(blocks.get(1), blocks.get(2));
+    assertThat(loop.hasNoWayOut()).isTrue();
+  }
+
+  @Test
+  public void embeddedMixedLoops() {
+    final CFG cfg = CFG.build(loader.getMethod("CFGLoopTest", "embeddedMixedLoops"));
+    Map<Tree, CFGLoop> loops = CFGLoop.getCFGLoops(cfg);
+    assertThat(loops).hasSize(2);
+    List<Block> blocks = CFGLoop.sorted(cfg.blocks());
+    CFGLoop loop = loops.get(blocks.get(11).terminator());
+    assertThat(loop.startingBlock()).isSameAs(blocks.get(11));
+    assertThat(loop.blocks()).containsOnly(blocks.get(10), blocks.get(9), blocks.get(4), blocks.get(3), blocks.get(2), blocks.get(1));
+    assertThat(loop.hasNoWayOut()).isTrue();
+    loop = loops.get(blocks.get(9).terminator());
+    assertThat(loop.startingBlock()).isSameAs(blocks.get(9));
+    assertThat(loop.blocks()).containsOnly(blocks.get(8), blocks.get(7), blocks.get(6), blocks.get(5));
+    assertThat(loop.hasNoWayOut()).isTrue();
+  }
+
+  @Test
+  public void mixedWithForEach() {
```

Comment:
This test should mention that `foreach` loops are simply not considered as being loops.